### PR TITLE
New Pool options and refactoring.

### PR DESCRIPTION
**Pull request by ifsnow**
I hope `Pool`'s features will get better like the others, so I added several options such as `queueWaitTimeout`, `pingCheckInterval`, `startConnections`, `minSpareConnections`, `maxSpareConnections`, `spareCheckInterval`
([See Readme.md](https://github.com/ifsnow/mysql/blob/improve-pool-options/Readme.md#pool-options))

## Efficient code
`Pool` will work 10~50 times faster with new options.
  
|  | New ([benchmark code](https://gist.github.com/ifsnow/c4672b09a12e2a34c05d3c5ffd9d2f1e)) | Original ([benchmark code](https://gist.github.com/ifsnow/609c393e007daaad5f49bd2c60d32509))|
| --- | --- | --- |
| Executions per second | 1,292,685 | 22,415 |
  
## Easy-to-read code
I've tried to make the code easier to read and maintain for the following contributors, and tidy up wrong and unused logics.

## Backwards compatibility code
All defaults for new options are off. All tests are passed, so existing users can use it right now.
In my opinion, It's recommended that they use the new options as other pools do.

Here is my config below.

```  
{
  connectionLimit : 50,
  queueWaitTimeout : 10000, // Same as acquireTimeout.
  pingCheckInterval : 10000, // The connection used in 10 seconds is reused without ping check.
  startConnections : 10, // 10 connections are created when the pool is started.
  minSpareConnections : 10, // 10 spare connections should be kept in the pool at all times.
  maxSpareConnections : 20, // No more than 20 spare connections.
  spareCheckInterval : 300000 // Check the spare connections every 5 minutes.
}
```

If you agree on my opinion, you can change the default value of new options. or, I think you could add a comment at Readme.md like this.
  
```
`pingCheckInterval`: ........ (Default: `0`, Recommendation: `10000`)
```

***

I hope this helps.
Please let me know if you have any concerns or advice.

### CODE REVIEW

**Comment by dougwilson** [comment]
Thanks a lot @ifsnow ! Is it possible to break this up into multiple PRs instead of one massive one? At the very lest to break it up into many different commits in the PR? It's a _lot_ to really look through and understand the edge cases that will break people if this is applied without a major version bump.

File: test/unit/pool/test-destroy-connection.js [hunk 1]
```diff
@@ -11,10 +11,10 @@ server.listen(common.fakeServerPort, function (err) {
   pool.getConnection(function (err, connection) {
     assert.ifError(err);
 
-    assert.strictEqual(connection, pool._allConnections[0]);
+    assert.strictEqual(connection, pool._connectionManager._allConnections._map[1]);
```
Comment (dougwilson) [inline test/unit/pool/test-destroy-connection.js]: Even though this is an underscore-prefixed property, people really do use it, as I've seen throughout Stackoverflow and issues here. We can't change this without a major at this point. Hopefully it's possible to implement the idel timeout stuff without making this change.

Comment (ifsnow) [inline test/unit/pool/test-destroy-connection.js]: It's a sad situation. I'll work to provide important old properties.

Comment (dougwilson) [inline test/unit/pool/test-destroy-connection.js]: At this point in the module, probably every single property is important in some way. I would err on the site of just not changing them if you don't really need to.

Comment (ifsnow) [inline test/unit/pool/test-destroy-connection.js]: I really understand. but, I think it's uncommon to use those properties such as `Pool._acquiringConnections` directly. It seems that it's not easy and efficient to work with all the existing ones.

I thought shortly.
```js
pool.createPool({
    ...,
    keepOldProperties : true
});
```
If the user set this, 4 properties(_acquiringConnections, _allConnections, _freeConnections, _connectionQueue) are updated. Isn't it enough?

Comment (ifsnow) [inline test/unit/pool/test-destroy-connection.js]: This PR does not make sense if keeping all properties is essential. I'm a little worried about what to do. Unfortunately, I think I'd better close this.

Comment (dougwilson) [inline test/unit/pool/test-destroy-connection.js]: @ifsnow I think a lot of the pool optimizations are worth it, and certainly can make it into a major version of this module. The only real concern I have is that because this is just one giant PR, making good (but breaking) changes _and_ making good but non-breaking changes, together, I have no option but to delay all the features in this PR until the next major. If you can simply break it up into the different pieces as multiple PRs, we can easily accept the easy ones now and discuss the breaking ones better and decide on the release of a new major version.

Comment (ifsnow) [inline test/unit/pool/test-destroy-connection.js]: @dougwilson I really know how difficult it's to review many changes. I'm going to think about how I can break it up but, it's not easy. That's because `PoolConnectionManager` has all new functions. I wonder if it's possible to split this into multiple PRs easily.
